sklearn/discriminant_analysis.py

```diff
@@ -427,7 +427,8 @@ def fit(self, X, y):
             Target values.
         """
         # FIXME: Future warning to be removed in 0.23
-        X, y = check_X_y(X, y, ensure_min_samples=2, estimator=self)
+        X = as_float_array(X)
```

Comment:
This line is necessary to get coefficients of type float32 output when X is of type int32. Removing this line gives coefficients of type float64.

Reply:
This is strange becasue the next `check_X_y` will convert to float anyway. It seems unecessary, isn't it?

Reply:
Oh I see why. So X will be converted to float64 and you expect it to be converted to float 32 because X is of type int32. I would say that there is no real reason for that. `int32` can default to `float64`. I am fine with that.

@massich what is the mechanism in the other estimators that you modified?

Reply:
There was no mechanism set. I've the vague idea that we talked about it long ago, but I can't recall that we agreed on anything. To me it makes sense that if someone willing creates the data in 32, wants to keep it in 32 bits. But if such individual is there she/he will scream at us, or would had done it already. So I've no objection on defaulting to `float64`.

Reply:
An int32 is a "very big" int. Hence, it make sense to convert it to a float64. People who want to save memory with ints use int16. Indeed, an int has a larger span than a float for the same memory cost.